File: python/src/ttio/importers/imported_dataset.py
"""``ImportedDataset`` — the normalized in-memory draft every importer
produces. The single call site of :meth:`SpectralDataset.write_minimal`,
collapsing the per-format adapter normalization the registry used to do.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING

from ..identification import Identification
from ..provenance import ProvenanceRecord
from ..quantification import Quantification
from ..spectral_dataset import SpectralDataset, WrittenRun

if TYPE_CHECKING:
    from ..written_genomic_run import WrittenGenomicRun


@dataclass(slots=True)
class ImportedDataset:
    """In-memory bundle of built runs + dataset metadata, ready to write."""

    title: str = ""
    isa_investigation_id: str = ""
    runs: dict[str, WrittenRun] = field(default_factory=dict)
    genomic_runs: dict[str, "WrittenGenomicRun"] = field(default_factory=dict)
    identifications: list[Identification] = field(default_factory=list)
    quantifications: list[Quantification] = field(default_factory=list)
    provenance: list[ProvenanceRecord] = field(default_factory=list)
    image: object | None = None
    raman_image: object | None = None
    ir_image: object | None = None
    subjects: list = field(default_factory=list)
    samples: list = field(default_factory=list)
    # Streaming sources (GenomicStreamSource / SpectralStreamSource):
    # written after write_minimal through the stream writers, so a run
    # of any size never sits in memory whole.
    genomic_streams: dict = field(default_factory=dict)
    spectral_streams: dict = field(default_factory=dict)

    def write(self, path: str | Path, *, features: list[str] | None = None,
              provider: str = "hdf5", progress=None) -> Path:
        out = self._write_static(path, features=features, provider=provider, progress=progress)
        if self.genomic_streams or self.spectral_streams:
            with SpectralDataset.open(out, writable=True, provider=provider) as ds:
                for src in self.genomic_streams.values():
                    src.write_into(ds.study_group, progress=progress)
                for src in self.spectral_streams.values():
                    src.write_into(ds.study_group, progress=progress)
        return out

    def _write_static(self, path: str | Path, *, features: list[str] | None = None,
                      provider: str = "hdf5", progress=None) -> Path:
        return SpectralDataset.write_minimal(
            path,
            title=self.title or "imported",
            isa_investigation_id=self.isa_investigation_id,
            runs=self.runs,
            genomic_runs=self.genomic_runs or None,
            identifications=self.identifications or None,
            quantifications=self.quantifications or None,
            provenance=self.provenance or None,
            features=features,
            provider=provider,
            image=self.image,
            raman_image=self.raman_image,
            ir_image=self.ir_image,
            subjects=self.subjects or None,
            samples=self.samples or None,
            progress=progress,
        )

File: java/src/main/java/global/thalion/ttio/importers/ImportedDataset.java
package global.thalion.ttio.importers;

import global.thalion.ttio.AcquisitionRun;
import global.thalion.ttio.Identification;
import global.thalion.ttio.IRImage;
import global.thalion.ttio.MSImage;
import global.thalion.ttio.ProvenanceRecord;
import global.thalion.ttio.Quantification;
import global.thalion.ttio.RamanImage;
import global.thalion.ttio.Sample;
import global.thalion.ttio.SpectralDataset;
import global.thalion.ttio.Subject;
import global.thalion.ttio.genomics.WrittenGenomicRun;
import global.thalion.ttio.io.ProgressSink;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Normalized in-memory draft every importer produces; {@link #write} is
 *  the single dataset-write call site. Cross-language equivalent: Python
 *  {@code ttio.importers.imported_dataset.ImportedDataset}. */
public final class ImportedDataset {
    public String title = "";
    public String isaInvestigationId = "";
    public final List<AcquisitionRun> runs = new ArrayList<>();
    public final List<WrittenGenomicRun> genomicRuns = new ArrayList<>();
    public final List<Identification> identifications = new ArrayList<>();
    public final List<Quantification> quantifications = new ArrayList<>();
    public final List<ProvenanceRecord> provenance = new ArrayList<>();
    public final List<Subject> subjects = new ArrayList<>();
    public final List<Sample> samples = new ArrayList<>();
    public MSImage image;
    public RamanImage ramanImage;
    public IRImage irImage;
    /** Genomic runs delivered as batch streams; written after the
     *  in-memory content through {@link GenomicStreamSource#writeInto}. */
    public final Map<String, GenomicStreamSource> genomicStreams = new LinkedHashMap<>();
    /** Spectral runs delivered as batch streams; written after the
     *  in-memory content through {@link SpectralStreamSource#writeInto}. */
    public final Map<String, SpectralStreamSource> spectralStreams = new LinkedHashMap<>();

    /** Optional write-through delegate. When non-null, {@link #write}
     *  routes the write through it instead of {@link SpectralDataset#create}.
     *  Used by CLI-delegated (write-through) importers — e.g.
     *  {@link BrukerTDFReader} — whose {@code .tio} is produced by an
     *  external process at write time, not assembled in memory. */
    @FunctionalInterface
    public interface WriteDelegate {
        java.nio.file.Path write(java.nio.file.Path output,
                global.thalion.ttio.io.ProgressSink progress)
                throws java.io.IOException;
    }

    /** When non-null, {@link #write} routes through this delegate. */
    public WriteDelegate writeDelegate;

    /** Create a delegate-backed draft for a write-through importer. */
    public static ImportedDataset delegated(WriteDelegate delegate) {
        ImportedDataset d = new ImportedDataset();
        d.writeDelegate = delegate;
        return d;
    }

    /** Write this draft to {@code output} with no progress reporting. */
    public Path write(Path output) throws java.io.IOException {
        return write(output, null);
    }

    /** Write this draft to {@code output} via the single image-aware
     *  {@link SpectralDataset#create} call site, optionally reporting
     *  per-section progress. Returns the written {@code .tio} path.
     *  When a {@link #writeDelegate} is set, the write is routed
     *  through it (write-through importers). */
    public Path write(Path output, ProgressSink progress) throws java.io.IOException {
        if (writeDelegate != null) {
            return writeDelegate.write(output, progress);
        }
        Path written = SpectralDataset.create(
            output.toString(),
            title.isEmpty() ? "imported" : title,
            isaInvestigationId,
            runs, genomicRuns,
            identifications, quantifications, provenance,
            subjects, samples,
            image, ramanImage, irImage,
            progress != null ? progress : ProgressSink.discard());
        if (!genomicStreams.isEmpty() || !spectralStreams.isEmpty()) {
            try (global.thalion.ttio.providers.StorageProvider p =
                     global.thalion.ttio.providers.ProviderRegistry.open(
                         written.toString(), global.thalion.ttio.providers.StorageProvider.Mode.READ_WRITE, "hdf5")) {
                global.thalion.ttio.providers.StorageGroup study = p.rootGroup().openGroup("study");
                for (GenomicStreamSource src : genomicStreams.values()) src.writeInto(study, progress);
                for (SpectralStreamSource src : spectralStreams.values()) src.writeInto(study, progress);
            }
        }
        return written;
    }
}
